### PR TITLE
Adding sibra service to manage steady paths.

### PR DESCRIPTION
**Pull request by kormat**
- Sibra service (SB) is responsible for setting up steady paths over any
  parent links, and for approving requests from any children.
- Beacon server sends up-segments to SB in addition to PS
- SB uses the same lib.sibra.state code as the router
- SB core and local modules exist, but are empty right now.
- SB currently has no redundancy, this will need further thought.
- Move SibraState tick updating into SibraState itself, as it's now used
  by multiple components.
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/618%23issuecomment-181370005%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52453946%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454370%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454440%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454712%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454753%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52455207%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52455285%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52456830%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52456862%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52476143%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52478366%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52479108%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23issuecomment-182451619%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52481027%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52481208%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23issuecomment-181370005%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40pszalach%20%40shitz%20%5Cr%5Cn%40perrig%20%40reischuk%20%22%2C%20%22created_at%22%3A%20%222016-02-08T13%3A34%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22LGTM%22%2C%20%22created_at%22%3A%20%222016-02-10T16%3A12%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/router/main.py%2052%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52453946%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%5C%22...to%20the%20SIBRA%20servers.%5C%22%20%3F%20rather%20%5C%22...to%20the%20SIBRA%20service.%5C%22%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A05%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Tweaked%20wording%20slightly.%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A14%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/router/main.py%3AL339-360%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/sibra_server/link.py%2051%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52455207%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22hm%2C%20actually%20what%20in%20the%20case%20of%20%5C%22ROUTING%5C%22%20links%3F%20WOuld%20it%20work%3F%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A18%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%5C%22ROUTING%5C%22%20links%20are%20for%20core%20paths%2C%20which%20aren%27t%20supported%20yet.%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A35%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/sibra_server/link.py%3AL1-164%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/sibra_server/link.py%2044%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454370%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22why%20sibra%20server%27s%20address%20belongs%20to%20this%20class%3F%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A10%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%27s%20needed%20when%20creating%20a%20packet%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A11%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22so%2C%20it%20should%20be%20passed%20then.%20Anyway%20it%20is%20minor...%22%2C%20%22created_at%22%3A%20%222016-02-10T16%3A08%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/sibra_server/link.py%3AL1-164%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/sibra_server/link.py%2046%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52454753%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22maybe%20%60assert%20iface.neighbor_type%20in%20%5B%5C%22PARENT%5C%22%2C%20%5C%22CHILD%5C%22%5D%60%20%3F%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A14%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%60assert%20iface.neighbor_type%20%21%3D%20%5C%22PEER%5C%22%60%20may%20be%20more%20accurate%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A19%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22There%20is%20a%20Link%28%29%20for%20every%20interface%2C%20no%20matter%20what%20the%20type.%20The%20non-parent/child%20ones%20don%27t%20do%20anything%20yet%2C%20but%20will%20once%20sibra%20supports%20peering%20links.%22%2C%20%22created_at%22%3A%20%222016-02-10T13%3A35%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/sibra_server/link.py%3AL1-164%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/sibra_server/util.py%2032%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52478366%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22last%3F%20it%20seems%20to%20return%20current%20ifid%22%2C%20%22created_at%22%3A%20%222016-02-10T16%3A03%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22They%20are%20the%20same%20in%20this%20case.%22%2C%20%22created_at%22%3A%20%222016-02-10T16%3A19%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/sibra_server/util.py%3AL1-51%22%7D%2C%20%22Pull%20e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47%20infrastructure/sibra_server/steady.py%2064%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/618%23discussion_r52476143%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22again...%20it%20really%20should%20not%20be%20here%22%2C%20%22created_at%22%3A%20%222016-02-10T15%3A51%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%27s%20used%20in%20%60_create_hdrs%60.%22%2C%20%22created_at%22%3A%20%222016-02-10T16%3A18%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/sibra_server/steady.py%3AL1-222%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#issuecomment-181370005'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> @pszalach @shitz
  @perrig @reischuk
- <a href='https://github.com/pszalach'></a> LGTM
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/router/main.py 52'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52453946'>File: infrastructure/router/main.py:L339-360</a></b>
- <a href='https://github.com/pszalach'></a> "...to the SIBRA servers." ? rather "...to the SIBRA service."
- <a href='https://github.com/kormat'></a> Tweaked wording slightly.
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/sibra_server/link.py 44'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52454370'>File: infrastructure/sibra_server/link.py:L1-164</a></b>
- <a href='https://github.com/pszalach'></a> why sibra server's address belongs to this class?
- <a href='https://github.com/kormat'></a> It's needed when creating a packet
- <a href='https://github.com/pszalach'></a> so, it should be passed then. Anyway it is minor...
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/sibra_server/link.py 46'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52454753'>File: infrastructure/sibra_server/link.py:L1-164</a></b>
- <a href='https://github.com/pszalach'></a> maybe `assert iface.neighbor_type in ["PARENT", "CHILD"]` ?
- <a href='https://github.com/pszalach'></a> `assert iface.neighbor_type != "PEER"` may be more accurate
- <a href='https://github.com/kormat'></a> There is a Link() for every interface, no matter what the type. The non-parent/child ones don't do anything yet, but will once sibra supports peering links.
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/sibra_server/link.py 51'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52455207'>File: infrastructure/sibra_server/link.py:L1-164</a></b>
- <a href='https://github.com/pszalach'></a> hm, actually what in the case of "ROUTING" links? WOuld it work?
- <a href='https://github.com/kormat'></a> "ROUTING" links are for core paths, which aren't supported yet.
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/sibra_server/steady.py 64'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52476143'>File: infrastructure/sibra_server/steady.py:L1-222</a></b>
- <a href='https://github.com/pszalach'></a> again... it really should not be here
- <a href='https://github.com/kormat'></a> It's used in `_create_hdrs`.
- [ ] <a href='#crh-comment-Pull e672e6556a9e2bebf2ee42bc4d28ef90ea2cbd47 infrastructure/sibra_server/util.py 32'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/618#discussion_r52478366'>File: infrastructure/sibra_server/util.py:L1-51</a></b>
- <a href='https://github.com/pszalach'></a> last? it seems to return current ifid
- <a href='https://github.com/kormat'></a> They are the same in this case.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/618?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/618?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/618'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: infrastructure/router/main.py [hunk 1]
```diff
@@ -342,6 +339,21 @@ def relay_cert_server_packet(self, spkt, from_local_ad):
             port = SCION_UDP_PORT
         self.send(spkt, addr, port)
 
+    def fwd_sibra_service_pkt(self, spkt, _):
+        """
+        Forward SIBRA service packets to the SIBRA servers.
```
Comment (pszal) [inline infrastructure/router/main.py]: "...to the SIBRA servers." ? rather "...to the SIBRA service."

File: infrastructure/sibra_server/link.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    """
+    def __init__(self, addr, sendq, iface):
+        """
+        :param ScionAddr addr: the address of this sibra server
+        :param queue.Queue sendq:
+            packets written to this queue will be sent by the sibra server
+            thread.
+        :param topology.InterfaceElement iface: the interface to manage.
+        """
+        self.addr = addr
```
Comment (pszal) [inline infrastructure/sibra_server/link.py:44]: why sibra server's address belongs to this class?

Comment (kormat) [inline infrastructure/sibra_server/link.py:44]: It's needed when creating a packet

File: infrastructure/router/main.py (hunk 1 above)
Comment (kormat) [inline infrastructure/router/main.py]: Tweaked wording slightly.

File: infrastructure/sibra_server/link.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        self.sendq = sendq
+        self.iface = iface
```
Comment (pszal) [inline infrastructure/sibra_server/link.py:46]: maybe `assert iface.neighbor_type in ["PARENT", "CHILD"]` ?

File: infrastructure/sibra_server/link.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        self.neigh = iface.isd_ad()
+        self.id = iface.if_id
+        self.state = SibraState(iface.bandwidth, self.addr.get_isd_ad())
+        self.segments = {}
+        self.parent = iface.neighbor_type == "PARENT"
```
Comment (pszal) [inline infrastructure/sibra_server/link.py:51]: hm, actually what in the case of "ROUTING" links? WOuld it work?

File: infrastructure/sibra_server/link.py (hunk 3 above)
Comment (pszal) [inline infrastructure/sibra_server/link.py:46]: `assert iface.neighbor_type != "PEER"` may be more accurate

Comment (kormat) [inline infrastructure/sibra_server/link.py:46]: There is a Link() for every interface, no matter what the type. The non-parent/child ones don't do anything yet, but will once sibra supports peering links.

File: infrastructure/sibra_server/link.py (hunk 4 above)
Comment (kormat) [inline infrastructure/sibra_server/link.py:51]: "ROUTING" links are for core paths, which aren't supported yet.

File: infrastructure/sibra_server/steady.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    def __init__(self, addr, sendq, seg, bwsnap):
+        """
+        :param ScionAddr addr: the address of this sibra server
+        :param queue.Queue sendq:
+            packets written to this queue will be sent by the sibra server
+            thread.
+        :param PathSegment seg: path segment to use.
+        :param BWSnapshot bwsnap: initial bandwidth to request.
+        """
+        self.addr = addr
```
Comment (pszal) [inline infrastructure/sibra_server/steady.py:64]: again... it really should not be here

File: infrastructure/sibra_server/util.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    tuples.
+    """
+    hops = []
+    # Ignore the last hop, as that's the local AD.
+    for ad in seg.ads[:-1]:
+        hops.append((ad.pcbm.isd_id, ad.pcbm.ad_id))
+    return tuple(hops)
+
+
+def find_last_ifid(pkt, ext):
```
Comment (pszal) [inline infrastructure/sibra_server/util.py:32]: last? it seems to return current ifid

File: infrastructure/sibra_server/link.py (hunk 2 above)
Comment (pszal) [inline infrastructure/sibra_server/link.py:44]: so, it should be passed then. Anyway it is minor...

File: infrastructure/sibra_server/steady.py (hunk 5 above)
Comment (kormat) [inline infrastructure/sibra_server/steady.py:64]: It's used in `_create_hdrs`.

File: infrastructure/sibra_server/util.py (hunk 6 above)
Comment (kormat) [inline infrastructure/sibra_server/util.py:32]: They are the same in this case.

**Pull request merged**